File: 2751-sliding-subarray-beauty/sliding-subarray-beauty.java
class Solution {
    public static int[] getSubarrayBeauty(int[] nums, int k, int x) {
        TreeMap<Integer, Integer> freqMap = new TreeMap<>();
        ArrayList<Integer> result = new ArrayList<>();

        int i = 0;
        int j = 0;

        while (j < nums.length) {
            if (nums[j] < 0) {
                freqMap.put(nums[j], freqMap.getOrDefault(nums[j], 0) + 1);
            }

            if (j - i + 1 == k) {
                int beauty = getKthSmallestNegative(freqMap, x);
                result.add(beauty);
                if (nums[i] < 0) {
                    int count = freqMap.get(nums[i]);
                    if (count == 1) {
                        freqMap.remove(nums[i]);
                    } else {
                        freqMap.put(nums[i], count - 1);
                    }
                }
                i++;
            }
            j++;
        }

        // Convert ArrayList to array
        int[] res = new int[result.size()];
        for (int idx = 0; idx < result.size(); idx++) {
            res[idx] = result.get(idx);
        }

        return res;
    }

    private static int getKthSmallestNegative(TreeMap<Integer, Integer> freqMap, int x) {
        if (freqMap.isEmpty()) {
            return 0; // If there are no negative numbers, return 0
        }

        int count = 0;
        for (Map.Entry<Integer, Integer> entry : freqMap.entrySet()) {
            count += entry.getValue();
            if (count >= x) {
                return entry.getKey(); // Return the xth smallest negative number
            }
        }
        return 0; // If there are fewer than x negative numbers, return 0
    }

}

File: 2751-sliding-subarray-beauty/sliding-subarray-beauty.cpp
class Solution {
public:
    vector<int> getSubarrayBeauty(vector<int>& nums, int k, int x) {
        int n = nums.size(),i = 0,j = 0,sz = 0;
        vector<int> ans;
        map<int,int> mp;
        while(j < n){
            if(nums[j] < 0){ 
                mp[nums[j]]++;
                sz++;
            }
            if(j-i+1 == k){
                if(sz >= x){
                    int ct = 0;
                    for(auto it:mp){
                        ct += it.second;
                        if(ct >= x){
                            ans.push_back(it.first);
                            break;
                        }
                    }
                }else{
                    ans.push_back(0);
                }
                if(nums[i] < 0){
                    mp[nums[i]]--;
                    if(mp[nums[i]] == 0) mp.erase(nums[i]);
                    sz--;
                }
                i++;
            }
			j++;
        }
        return ans;
    }
};